Operations Documentation (#6)
EGI-Foundation/notebooks: source/operations.rst

Service Operations
------------------

In this section you can find the common operational activities related to keep
the service available to our users.

Initial set-up
==============

Notebooks VO
::::::::::::

The resources used for the Notebooks deployments are managed with the 
``vo.notebooks.egi.eu`` VO. Operators of the service should join the VO, check
the entry at the `operations portal <https://operations-portal.egi.eu/vo/view/voname/vo.notebooks.egi.eu>`_
and at `AppDB <https://appdb.egi.eu/store/vo/vo.notebooks.egi.eu>`_.

Clients installation
::::::::::::::::::::

In order to manage the resources you will need these tools installed
on your client machine:

* ``egicli`` for discovering sites and managing tokens,

* ``terraform`` to create the VMs at the providers,

* ``ansible`` to configure the VMs and install kubernetes at the providers,

* ``terraform-inventory`` to get the list of hosts to use from terraform.

> 

Reviewer: ```suggestion
```

Get the configuration repo
::::::::::::::::::::::::::

All the configuration of the notebooks is stored at a git repo available in 

> keybase, start by cloning the repo:

Reviewer: should probably document that membership to a special team is required?

Author: Yes, will update.

Anyway, I'm not super-convinced of the keybase approach, would rather have secrets only stored somewhere (which are very few things, 2-3 keys and the check-in clients) and then have the config in GitHub as we could do PR based changes

Reviewer: yes Keybase is a temporary solution/workaround before having a cleaner way of doing this

.. code-block:: shell
   
   $ git clone keybase://team/opslife/egi-notebooks

> 

Reviewer: ```suggestion
```

Kubernetes
==========

We use ``terraform`` and ``ansible`` to build the cluster at one of the EGI Cloud
providers. If you are building the cluster for the first time, create a new

> directory on your local git repository from the templat, add it to the

Reviewer: ```suggestion
directory on your local git repository from the template, add it to the
```

repo, and get ``terraform`` ready:

.. code-block:: shell

   $ cp -a template <new provider>
   $ git add <new provider>
   $ cd <new provider>/terraform
   $ terraform init

Using the ``egicli`` you can get the list of projects and their ids
for a given site:

.. code-block:: shell

   $ egicli endpoint projects --site CESGA
   id                                Name                 enabled    site
   --------------------------------  -------------------  ---------  ------
   3a8e9d966e644405bf19b536adf7743d  vo.access.egi.eu     True       CESGA
   916506ac136741c28e4326975eef0bff  vo.emso-eric.eu      True       CESGA
   b1d2ef2cc2284c57bcde21cf4ab141e3  vo.nextgeoss.eu      True       CESGA
   eb7ff20e603d471cb731bdb83a95a2b5  fedcloud.egi.eu      True       CESGA
   fcaf23d103c1485694e7494a59ee5f09  vo.notebooks.egi.eu  True       CESGA

> And with the project id, you can obtain all the environment variables needed

Reviewer: ```suggestion
And with the project ID, you can obtain all the environment variables needed
```

to interact with the OpenStack APIs of the site:

.. code-block:: shell

   $ eval "$(egicli endpoint env --site CESGA --project-id fcaf23d103c1485694e7494a59ee5f09)"

Now you are ready to use the openstack or terraform at the site. The token
obtained is valid for 1 hour, you can refresh it at any time with:

.. code-block:: shell

   $ eval "$(egicli endpoint token --site CESGA --project-id fcaf23d103c1485694e7494a59ee5f09)"

> 

Reviewer: ```suggestion
```

> First get the network ids and pool to use for the site:

Reviewer: ```suggestion
First get the network IDs and pool to use for the site:
```

.. code-block:: shell

   $ openstack network list
   +--------------------------------------+-------------------------+--------------------------------------+
   | ID                                   | Name                    | Subnets                              |
   +--------------------------------------+-------------------------+--------------------------------------+
   | 1aaf20b6-47a1-47ef-972e-7b36872f678f | net-vo.notebooks.egi.eu | 6465a327-c261-4391-a0f5-d503cc2d43d3 |
   | 6174db12-932f-4ee3-bb3e-7a0ca070d8f2 | public00                | 6af8c4f3-8e2e-405d-adea-c0b374c5bd99 |
   +--------------------------------------+-------------------------+--------------------------------------+

> 

Reviewer: ```suggestion
```

In this case we will use ``public00`` as the pool for public IPs and 

> ``1aaf20b6-47a1-47ef-972e-7b36872f678f`` as the network id. Check with the provider

Reviewer: ```suggestion
``1aaf20b6-47a1-47ef-972e-7b36872f678f`` as the network ID. Check with the provider
```

which is the right network to use. Use these values in the ``terraform.tfvars``
file:

.. code-block:: terraform

   ip_pool = "public00"
   net_id  = "1aaf20b6-47a1-47ef-972e-7b36872f678f"

> 

Reviewer: ```suggestion
```

   | 6ba0080d-d71c-4aff-b6f9-b5a9484097f8 | small          |   512 |    2 |         0 |     1 | True      |
   | 6e514065-9013-4ce1-908a-0dcc173125e4 | cor2mem4hd20   |  4096 |   20 |         0 |     2 | True      |
   | 85f66ce6-0b66-4889-a0bf-df8dc23ee540 | cor1mem2hd10   |  2048 |   10 |         0 |     1 | True      |
   | c4aa496b-4684-4a86-bd7f-3a67c04b1fa6 | cor24mem50hd50 | 51200 |   50 |         0 |    24 | True      |
   | edac68c3-50ea-42c2-ae1d-76b8beb306b5 | test-bigHD     |  4096 |  237 |         0 |     2 | True      |
   +--------------------------------------+----------------+-------+------+-----------+-------+-----------+

Finally ensure your public ssh key is also listed in the ``cloud-init.yaml``
file and then you are ready to deploy the cluster with:

.. code-block:: shell

   $ terraform apply

Your VMs are up and running, it's time to get kubernetes configured and running
with ansible:

.. code-block:: shell
  
   $ cd ..   # you should be now in <new provider>
   $ ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS=silently TF_STATE=./terraform \
     ansible-playbook --inventory-file=$(which terraform-inventory) \
     playbooks/k8s.yaml

> 

Reviewer: ```suggestion
```

Interacting with the cluster
::::::::::::::::::::::::::::

As the master will be on a private IP, you won't be able to directly interact

> with it, but you can still ssh into the VM using the ingress node as a bastion

Reviewer: Not sure bastion will mean something to all maybe something like gateway or relay will be more clear?

   cluster-ingress	3       	Wed Jan  8 15:56:53 2020	DEPLOYED	nginx-ingress-1.7.0         	0.24.1     	kube-system 
   nfs-provisioner	3       	Wed Jan  8 15:56:43 2020	DEPLOYED	nfs-client-provisioner-1.2.8	3.1.0      	kube-system 

Modifying/Destroying the cluster
::::::::::::::::::::::::::::::::

You should be able to change the number of workers in the cluster and re-apply
terraform to start them and then execute the playbook to get them added to the
cluster.

Any changes in the master, NFS or ingress VMs should be done carfully as those
will probably break the configuration of the kubernetes cluster and of any
application running on top.

.. TODO: remove nodes?

.. TODO: update master/ingress/nfs

Destroying the cluster can be done with a single command:

.. code-block:: shell

   $ terraform destroy

Notebooks deployments
=====================

Once the k8s cluster is up and running, you can deploy a notebooks instance.
For each deployment you should create a file in the `deployments` directory
following the template provided:

.. code-block:: shell

   $ cp deployments/hub.yaml.template deployments/hub.yaml

>    

Reviewer: ```suggestion
```

> Each deployment will need a a domain name pointing to your ingress host, you

Reviewer: ```suggestion
Each deployment will need a domain name pointing to your ingress host, you
```

can create one at the `FedCloud dynamic DNS service <https://nsupdate.fedcloud.eu/>`_.

> Then you will need to create a client for EGI Check-in to authorise users

Reviewer: ```suggestion
Then you will need to create an OpenID Connect client for EGI Check-in to authorise users
```

into the new deployment. You can create a client by going to the `Check-in demo
OIDC clients management <https://aai-demo.egi.eu/oidc/manage/admin/clients>`_.

> Use the followin as redirect URL: ``https://<your host domain name>/hub/oauth_callback``.

Reviewer: ```suggestion
Use the following as redirect URL: ``https://<your host domain name>/hub/oauth_callback``.
```

In the `Access` tab, add ``offline_access`` to the list of scopes. Save the

> client and take note of the client id and client secret for later.

Reviewer: ```suggestion
client and take note of the client ID and client secret for later.
```

Finally you will also need 3 different random strings generated with
``openssl rand -hex 32`` that will be used as secrets in the file describing 
the deployment. 

Go and edit the deployment description file to add this information (search for
``# FIXME NEEDS INPUT`` in the file to quickly get there)

For deploying the notebooks instance we will also use ``ansible``:

.. code-block:: shell

   $ ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS=silently TF_STATE=./terraform ansible-playbook \
          --inventory-file=$(which terraform-inventory) playbooks/notebooks.yaml
   
The first deployment trial may fail due to a timeout caused by the downloading

> of the container images needed. You can retry after a while to re-deploy.

Reviewer: Are the ansible playbooks idempotent?

Author: yep :)

ansible here is mostly used as a way to avoid having access to the k8s master with public IP or sshing through a bastion. The playbook will upgrade the helm of the notebooks and that will not make any difference to what's deployed if the description of the service does not change. A new upgrade will be registered to the deployment (but I think is harmless to have the count increased +1)

Reviewer: Perfect, closing that comment.

Author: The k8s setup playbook seems to be also idempotent, so you can add nodes easily without breaking stuff. That said, changing the master, the NFS or the ingress will cause disruptions

   ==> v1/StatefulSet
   NAME              DESIRED  CURRENT  AGE
   user-placeholder  0        0        6m44s

   ==> v1beta1/Ingress
   NAME        HOSTS                                 ADDRESS  PORTS  AGE
   jupyterhub  notebooktest.fedcloud-tf.fedcloud.eu  80, 443  6m44s

   ==> v1beta1/PodDisruptionBudget
   NAME              MIN AVAILABLE  MAX UNAVAILABLE  ALLOWED DISRUPTIONS  AGE
   hub               1              N/A              0                    6m48s
   proxy             1              N/A              0                    6m48s
   user-placeholder  0              N/A              0                    6m48s
   user-scheduler    1              N/A              1                    6m47s

   ==> v1/Role
   NAME  AGE
   hub   6m46s

   NOTES:
   Thank you for installing JupyterHub!

   Your release is named hub and installed into the namespace hub.

   You can find if the hub and proxy is ready by doing:

   kubectl --namespace=hub get pod

   and watching for both those pods to be in status 'Running'.

   You can find the public IP of the JupyterHub by doing:

   kubectl --namespace=hub get svc proxy-public

   It might take a few minutes for it to appear!

>    Note that this is still an alpha release! If you have questions, feel free to

Reviewer: Aren't we at Beta now? :)

Author: That's jupyterhub helm notes, we are production :)

Reviewer: OK cool :)
